### ISSUE TITLE
`cargo` and `rustup` get deleted when saving cache

### ISSUE THREAD

**Issue by Procrat** [Swatinem/rust-cache#16]
Since we switched to running on self-hosted runners, we noticed that some jobs failed because they couldn't find the `cargo` executable anymore. After some digging, we noticed that after a successful run with this action on the same machine, `.cargo/bin` was empty, which is normally where `cargo` and `rustup` are installed (when installing the [rustup.rs](https://rustup.rs/) way at least).

Looking through the code, it seems that `.cargo/bin` gets cleaned apart for the binaries it finds a matching crate for, which would end up removing `cargo` and `rustup`. Was this intentional? If so, should I be doing something differently?

**Comment by Swatinem** [comment]
Yes that was intentional. Since `.cargo/bin` is saved in the cache as well, I cleared out all the not directly installed binaries from there.

Under the assumption that runners get a fresh snapshot every time they run, it was a safe thing to do.

**Comment by ErichDonGubler** [comment]
> Under the assumption that runners get a fresh snapshot every time they run, it was a safe thing to do.

This is not a safe assumption, I think? We at @NZXTCorp are interested in giving this action a try, but currently our self-hosted runners just call `rustup show` to force installation of overridden toolchain in our repos, This breaks on self-hosted runners because...well, `rustup.exe` is assumed to exist in the first place!

**Comment by Swatinem** [comment]
I think a reasonable workaround in that case is to move the files that are not supposed to be cached away, and back after saving the cache.
Or alternatively to use a dedicated folder just for the bin files that are supposed to be persisted in the cache, and move them to `.cargo/bin` on "startup".

**Comment by fpoli** [comment]
Instead of an option to not touch `.cargo/bin`, I would consider modifying this action so that it *temporarly* moves the unwanted files away instead of removing them. The code is around here:

https://github.com/Swatinem/rust-cache/blob/b8e72aae8381e4b9942ff10e8ec0fcd359dc5246/src/cleanup.ts#L72-L86

**Comment by Swatinem** [comment]
I had the exact same idea:
- temporarily move files out of the directory
- persist the cache so it uploads only the currently existing files
- move files back afterwards

I just never put in the effort to make that happen ;-) But feel free to contribute here. :-)

**Comment by stefnotch** [comment]
Would it be possible to change the cache paths ( https://github.com/Swatinem/rust-cache/blob/d30f1144e884b1fe579845ed7a339c640dff522e/src/config.ts#L237 ) to be more granular, and to simply ignore the `.cargo/bin` folder?
With more granular paths, we should be able to get rid of all of those deletion commands
https://github.com/Swatinem/rust-cache/blob/d30f1144e884b1fe579845ed7a339c640dff522e/src/save.ts#L49-L69
